audioipc/src/ipccore.rs

```diff
@@ -0,0 +1,717 @@
+                            Ok(done) => assert!(!done),
+                            Err(e) => {
+                                error!("[{:?}]: connection error: {:?}", token, e);
+                            }
+                        }
+                    } else {
+                        debug!("[{:?}]: token not found in slab: wake_connection", token);
+                        debug_assert!(false); // This shouldn't happen, catch it in debug mode.
+                    };
+                }
+            }
+        }
+
+        Ok(true)
+    }
+}
+
+// Connection wraps an interprocess connection (Pipe) and manages
+// receiving inbound and sending outbound buffers (and associated handles, if any).
+// The associated driver is responsible for message framing and serialization.
+struct Connection {
+    io: sys::Pipe,
+    token: Token,
+    interest: Interest,
+    inbound: sys::ConnectionBuffer,
+    outbound: sys::ConnectionBuffer,
+    driver: Box<dyn Driver + Send>,
+}
+
+const IPC_CLIENT_BUFFER_SIZE: usize = 16384;
+
+impl Connection {
+    fn new(
+        mut io: sys::Pipe,
+        token: Token,
+        driver: Box<dyn Driver + Send>,
+        registry: &Registry,
+    ) -> Result<Connection> {
+        let interest = Interest::READABLE;
+        registry.register(&mut io, token, interest)?;
+        Ok(Connection {
+            io,
+            token,
+            interest,
+            inbound: sys::ConnectionBuffer::with_capacity(IPC_CLIENT_BUFFER_SIZE),
+            outbound: sys::ConnectionBuffer::with_capacity(IPC_CLIENT_BUFFER_SIZE),
+            driver,
+        })
+    }
+
+    fn shutdown(mut self, registry: &Registry) -> Result<()> {
+        registry.deregister(&mut self.io)
+    }
+
+    // Update connection registration with the current readiness event interests.
+    fn update_registration(&mut self, registry: &Registry) -> Result<()> {
+        registry.reregister(&mut self.io, self.token, self.interest)
+    }
+
+    // Handle readiness event.  Errors returned are fatal for this connection, resulting in removal from the EventLoop connection list.
+    // The EventLoop will call this for any connection that has received an event, including external wake events to clear the outbound buffer.
+    fn handle_event(&mut self, event: Option<&Event>, registry: &Registry) -> Result<bool> {
+        if let Some(event) = event {
+            assert_eq!(self.token, event.token());
+        }
+        debug!("[{:?}]: handling event {:?}", self.token, event);
+
+        // If the connection is readable, read into inbound and pass to driver for processing until all ready data
+        // has been consumed.
+        //let done = if let Some(event) = event {
+        let done = if let Some(event) = event {
+            if event.is_readable() {
+                loop {
+                    trace!("[{:?}]: pre-recv inbound: {:?}", self.token, self.inbound);
+                    let r = self.io.recv_msg(&mut self.inbound);
```

Comment:
`IPC_CLIENT_BUFFER_SIZE` is an arbitrary starting buffer size, it's not expected to be large enough for all possible messages.  When `decode` is called, it will always ensure there's sufficient space available to read the next message length prefix (at message start) or next entire message (once message length is known from the length prefix).  The EventLoop will continue reading and attempting to decode as long as there's data available on the connection, causing `decode` to adjust the buffer size as necessary.

It'd be better if the buffer size management was clearer, I'll think about how it might be improved.